caffe2/operators/onnxifi_op.cc

```diff
@@ -0,0 +1,192 @@
+    tensor_descriptor.dataType = ONNXIFI_DATATYPE_FLOAT32;
+    tensor_descriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+    tensor_descriptor.dimensions = tensor_dims.size();
+    input_shapes_.emplace_back();
+    auto& input_shape = input_shapes_.back();
+    for (unsigned j = 0U; j < tensor_descriptor.dimensions; ++j) {
+      input_shape.push_back(tensor_dims[j]);
+    }
+    tensor_descriptor.shape = input_shape.data();
+    tensor_descriptor.buffer = (onnxPointer)(input_tensor.data<float>());
+  }
+
+  for (unsigned i = 0U; i < OutputSize(); ++i) {
+    auto* output_tensor = Output(i);
+    std::vector<TIndex> tensor_dims;
+    SetOutputShape(i, &tensor_dims);
+    output_tensor->Resize(tensor_dims);
+    auto& tensor_descriptor = output_desc_.at(i);
+    tensor_descriptor.dataType = ONNXIFI_DATATYPE_FLOAT32;
+    tensor_descriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+    tensor_descriptor.dimensions = tensor_dims.size();
+    output_shapes_.emplace_back();
+    auto& output_shape = output_shapes_.back();
+    for (unsigned j = 0U; j < tensor_descriptor.dimensions; ++j) {
+      output_shape.push_back(tensor_dims[j]);
+    }
+    tensor_descriptor.shape = output_shape.data();
+    tensor_descriptor.buffer =
+        (onnxPointer)(output_tensor->mutable_data<float>());
+  }
+
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxSetGraphIO(
+          graph_,
+          input_desc_.size(),
+          input_desc_.data(),
+          output_desc_.size(),
+          output_desc_.data()),
+      ONNXIFI_STATUS_SUCCESS);
+
+  onnxMemoryFence input_fence;
+  input_fence.event = nullptr;
+  input_fence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxInitEvent(backend_, input_fence.event), ONNXIFI_STATUS_SUCCESS);
+  onnxMemoryFence output_fence;
+  output_fence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
+  output_fence.event = nullptr;
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxInitEvent(backend_, output_fence.event), ONNXIFI_STATUS_SUCCESS);
+
+  // Singal event on input fence, call the asycn run on backend and wait for the event on output fence
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxSignalEvent(input_fence.event), ONNXIFI_STATUS_SUCCESS);
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxRunGraph(graph_, &input_fence, &output_fence), ONNXIFI_STATUS_SUCCESS);
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxWaitEvent(output_fence.event), ONNXIFI_STATUS_SUCCESS);
+
+  // Destroy the event objects
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxReleaseEvent(input_fence.event), ONNXIFI_STATUS_SUCCESS);
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxReleaseEvent(output_fence.event), ONNXIFI_STATUS_SUCCESS);
+
+  return true;
+}
+
+REGISTER_CPU_OPERATOR(Onnxifi, OnnxifiOp<float, CPUContext>);
+OPERATOR_SCHEMA(Onnxifi)
+    .NumInputs(0, INT_MAX)
+    .NumOutputs(0, INT_MAX)
+    .SetDoc(R"DOC(
+    The Onnxifi operator is a black-box operator to lower the computation to Onnxifi backend
+    )DOC")
+    .Arg("onnxifi_backend", "(string default=\"\") Name of the backend")
```

Comment:
We don't need this argument, it duplicates https://github.com/pytorch/pytorch/blob/master/caffe2/proto/caffe2.proto#L162

Reply:
Which field in `DeviceOption` should I encode this info? `extra_info`?

Reply:
The resolution from discussion with @bddppq and @pietern is that we should add new optional properties (which reflect the properties in `onnxGetBackendInfo`) in `DeviceOption` and match them to ONNXIFI backend at runtime.